File: project-euler-c++/problem64.cpp
// Project Euler Problem 64 (https://projecteuler.net/problem=64)
// July 20, 2020

#include<vector>
#include <math.h>	// floor, sqrt
#include "auxFunctions.h"	// checkPerfectSquare, findContinuedFraction
using namespace std;

int oddPeriodSquareRoots(int bound) {
	// returns the number of continued fractions of root(n) that have an odd period for n <= bound
	
	int oddCount = 0;

	// finding continued fractions with odd period
	for (int N = 2; N <= bound; N++)
		if (!aux::checkPerfectSquare(N) && aux::findContinuedFraction(N).size() % 2 == 1)
			oddCount++;

	return oddCount;
}

File: project-euler-julia/problem64.jl
# Project Euler Problem 64 (https://projecteuler.net/problem=64)
# July 20, 2020
# Runtime: 10⁻² seconds

import AuxFunctions: findContinuedFraction

"""
    oddPeriodSquareRoots(bound)

Returns the number of continued fractions of `√n` for `n ≤ bound` that have an odd period. 
"""
function oddPeriodSquareRoots(bound::Integer)
	return count(N -> length(findContinuedFraction(N)) % 2 == 1, 1:bound)
end

# function call
@btime oddPeriodSquareRoots(10000)